### ISSUE TITLE
as_tf_dataset: add shuffle=/seed= to shuffle the RID list (fixes class-grouped-bag training collapse)

### ISSUE THREAD

**Issue by carlkesselman** [informatics-isi-edu/deriva-ml#362]
## Summary

Add optional `shuffle: bool = False` and `seed: int | None = None` parameters to
`DatasetBag.as_tf_dataset` (and the underlying `build_tf_dataset` in
`deriva_ml/dataset/tf_adapter.py`) that shuffle the **RID list** before the
generator iterates it.

## Motivation

`as_tf_dataset` iterates `rids` in the order they come out of the bag. When a
bag's members are grouped by class — which happens for datasets assembled by
appending one class at a time — every batch drawn from the head of the
un-`.shuffle()`'d stream is single-class. For binary VGG-19 glaucoma training in
`eye-ai-vgg19` this reproducibly collapsed the model: `val_acc` oscillated
between the two class proportions (0.623 ↔ 0.377) and val AUC sat at ~0.5.

Callers can't cheaply fix this downstream:

- `tf.data.Dataset.shuffle(buffer_size)` on the built dataset only approximates a
  global shuffle unless `buffer_size >= len(dataset)`, and a full-size buffer
  means holding **every decoded image** in memory — infeasible for image bags.
- Shuffling the RID list **before** the generator runs is a global shuffle at
  effectively zero memory cost: it reorders a list of short strings, and images
  are still decoded lazily one at a time inside the generator.

So the correct fix belongs where the RID list is materialized — inside
`build_tf_dataset` — not in caller code.

## Proposed change

In `deriva_ml/dataset/tf_adapter.py`, `build_tf_dataset`, right after the `rids`
list is finalized (the `if targets is None: rids = all_rids …` block) and
**before** `_generate` closes over it:

```python
if shuffle:
    # Global, memory-free shuffle: reorder the RID list (short strings) in
    # place; images are still decoded lazily inside _generate one at a time.
    # Use an explicit local RNG, never the global `random` module — data
    # ordering must be reproducible from `seed` alone and independent of any
    # other random.* call elsewhere in the process.
    import random
    rids = list(rids)  # copy: don't mutate all_rids / a caller-visible list
    random.Random(seed).shuffle(rids)
```

Signature additions (keyword-only, defaults preserve current behavior) on both
`build_tf_dataset` and the `as_tf_dataset` wrapper that forwards to it:

```python
shuffle: bool = False,
seed: int | None = None,
```

and forward `shuffle=shuffle, seed=seed` in `DatasetBag.as_tf_dataset`'s call to
`build_tf_dataset`.

### Invariants preserved

- **RID-is-always-last** ordering of yielded tuples is untouched — only the
  order of `rids` changes, not the per-element tuple shape.
- **Default is a no-op**: `shuffle=False` iterates `rids` exactly as today.
- **`seed=None`** is a valid, documented choice (nondeterministic shuffle);
  reproducible callers pass an explicit int.
- **Interaction with `output_signature=None`**: the eager `next(gen)` used to
  infer the signature draws the *first shuffled* element, which is fine — the
  re-wrapping generator (`_generate_with_first`) already preserves it.

## Reproducibility note (for callers)

A single build-time shuffle produces the **same order every epoch** — fully
reproducible and sufficient to break the class-grouping collapse. If per-epoch
reshuffling is ever wanted, that's a separate follow-up (e.g. derive
`seed + epoch`); it should not change the semantics proposed here.

The `eye-ai-vgg19` caller records this seed as a first-class `shuffle_seed`
hyperparameter so the training-batch composition is recoverable from the
execution's `run_config.json` and from standalone inference bundles.

## Downstream caller (already implemented, guarded)

`eye-ai-vgg19` `build_image_dataset` forwards `shuffle=True, seed=shuffle_seed`
for the **training** role only (validation/testing stay deterministically
ordered so recorded per-RID predictions are reproducible). It currently guards
with a `TypeError` catch that fails **loud** ("refusing to train unshuffled")
until this parameter lands, rather than silently training on a class-grouped
bag.

---
Filed from eye-ai-vgg19; context in that repo's `tacit-knowledge.md` (tk-036 root
cause, tk-037 this fix, tk-038 seed-as-recorded-hyperparameter).

**Comment by carlkesselman** [comment]
Implemented in #363, with one amendment to the proposed API.

**Parameter names: `global_shuffle=` / `shuffle_seed=`, not `shuffle=` / `seed=`.**

`shuffle` / `seed` collide with `tf.data.Dataset.shuffle(buffer_size, seed=...)`, which a TF user already knows and which behaves differently on the axis they care about: `.shuffle()` reshuffles every epoch by default, while ours is one build-time permutation, identical every epoch. Same word, opposite epoch semantics — and the mismatch is silent, since training converges and looks plausible while epoch 7 replays epoch 1's order.

The second-order trap is the real cost. A user who sees `shuffle=True` in the constructor concludes shuffling is handled and **drops `.shuffle()` from their pipeline**, trading a per-epoch shuffle for a fixed one without being told. The parameter's presence invites removing the thing doing the more useful work. `seed` compounds it: in `tf.data` it seeds the reservoir, in ours a `random.Random` over the RID list — two RNGs, two scopes, one name.

`global_shuffle` names the **guarantee** (global vs. windowed) rather than the mechanism. That's the axis that actually differs from `.shuffle()`, it reads coherently sitting beside `.shuffle(buffer_size=1000)`, and it stays accurate if we ever permute something other than the RID list. ("RIDs vs. samples" was the other candidate, but that's deriva-ml-internal vocabulary a TF user doesn't carry.)

No name can carry the epoch semantics, so the docstring's first line states them outright, and the docs recommend **both** knobs with the division of labor spelled out: `global_shuffle` fixes global class grouping the buffer can't reach, `.shuffle()` adds per-epoch variation. Full reproducibility needs both seeds — worth noting for the `run_config.json` provenance case, since recording `shuffle_seed` alone only pins batch composition if the rest of the pipeline is deterministic.

Two smaller deviations from the proposed patch:

- `import random` at module scope rather than inside the `if` block. It's stdlib; the lazy imports in that module are deliberately reserved for the optional torch/tf deps.
- No passthrough for `.shuffle()`'s own seed. That call happens on the returned dataset in user code where `seed=` is already reachable; accepting it here would mean also accepting `buffer_size`, `reshuffle_each_iteration`, and the rest — wrapping the `tf.data` builder API one parameter at a time.

Everything else follows the proposal as written: keyword-only, `False` default is a no-op, copy-before-shuffle, explicit `random.Random(seed)` never the global module, RID-is-always-last preserved, and the eager `output_signature` inference survives the reorder.

**On the downstream guard** — `eye-ai-vgg19`'s `TypeError` catch fails loud on the *old* kwarg name, so the rename surfaces immediately rather than silently no-op'ing into unshuffled training. `build_image_dataset` needs updating to `global_shuffle=True, shuffle_seed=shuffle_seed` when the pin advances.

Also worth confirming your reasoning on scope, since #363 keeps it TF-only: `as_torch_dataset` is map-style, so `DataLoader(shuffle=True)` already gives a global *and* per-epoch shuffle by permuting indices — an adapter-level flag there would actively regress anyone who set only it. And `restructure_assets`'s class-grouped tree is the contract `ImageFolder` reads labels from, so there's nothing to shuffle. TF is the one path where the framework can't rescue the caller.
